libs/constrained-generators/src/Constrained/GenT.hs

```diff
@@ -224,6 +226,20 @@ suchThatT g p = do
     go n cont = do
       a <- g
       if p a then pure a else scaleT (+ 1) $ go (n - 1) cont
+-}
+
+suchThatWithTryT :: MonadGenError m => Int -> GenT m a -> (a -> Bool) -> GenT m a
+suchThatWithTryT trys g p = do
+  mode <- getMode
+  let (n, cont) = case mode of
+        Strict -> (trys, fatalError)
+        Loose -> (1 :: Int, genError) -- TODO: Maybe 1 is not the right number here!
+  go n cont
+  where
+    go 0 cont = cont (pure ("Ran out of tries (" ++ show trys ++ ") on suchThatWithTryT"))
+    go n cont = do
+      a <- g
+      if p a then pure a else scaleT (+ 1) $ go (n - 1) cont
```

Comment:
```suggestion
      if p a then g else scaleT (+ 1) $ go (n - 1) cont
```

Reply:
I think this is clearer. So I am going to leave it as it is.